Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1330 @@
+
+Here we use the `many` function, which is very similar to the `some` function discussed before. The difference is that `any` tries to run the given parser **0 or more times**. In Regex, our `identifier` parser is equivalent to `[a-zA-Z][a-zA-Z0-9]*`.
+
+```hs
+identifier :: Parser Identifier
+identifier = label "identifier" $ do
+  first <- letterChar <|> char '_'
+  rest <- many $ alphaNumChar <|> char '_'
+  pure $ Identifier $ first : rest
+-- Or:
+-- identifier = Identifier <$> label "identifier" ((:) <$> (letterChar <|> char '_') <*> many (alphaNumChar <|> char '_'))
+```
+
+Now include `SId <$> identifier` to `atom`.
+
+**Warning**: When parsing with `<|>` or `choice`, **the order of the parsers matters**. Take this as an example:
+
+```hs
+>>> parseTest (SBool <$> bool <|> SId <$> identifier) "true"
+SBool True
+
+>>> parseTest (SId <$> identifier <|> SBool <$> bool) "true"
+SId (Identifier {getId = "true"})
+```
+
+In the first case, we parse a boolean before an identifier, which causes `true` to be matched by `bool`. However, in the second example, using `identifier` before `bool` caused `true` to be recognized as an identifier. You may assume that parsers are greedy, and you should be careful while considering such things. Notice that in the second case, the `bool` parser will never be executed.
+
+In other words, make sure that you parse a boolean in `atom` before you parse an identifier.
+
+### S-expressions
+
+Finally, let's move on to our last parser, which parses S-expressions themselves, such as `(foo 42 "hey")` or `(bar)`.
+
+First, we will do it the naïve way, which may come as the first natural solution to someone just learning Megaparsec, and then we will do it in a better manner using extra tools offered by Megaparsec.
+
+#### S-expressions: The awkward way
+
+The idea here is to parse an atom, followed by 0 or more atoms. Of course, this means that we want to use `many` to parse such atoms. Additionally, an S-expression may be surrounded by parenthesis, so we can use `between` for such a task.
+
+This parser, however, is not as simple as before, but let's first do it the wrong way (which won't work properly), analyze the problem, and come up with a solution to it.
+
+```hs
+sexp :: Parser (SExp, [SExp])
+sexp = label "S-expression" $ between (char '(') (char ')') ((,) <$> atom <*> many atom)
+-- Or:
+-- sexp = label "S-expression" $ char '(' *> ((,) <$> atom <*> many atom) <* char ')'
+```
+
+After adding it to `atom` with `uncurry SSExp <$> sexp`, we can test it in GHCi:
+
+```hs
+>>> parseTest atom "(foo)"
+SSExp (SId (Identifier {getId = "foo"})) []
+```
+
+Ok, looks good so far. What if we have arguments being applied to it?
```

Comment:
Isn't it other way around: what if we apply it to arguments?

Reply:
I don't think it is... I meant we have arguments applied to the function. But I reworded it in any case, hopefully, it's a bit more clear now. Or do we apply a function to its arguments? I'm now confused on what is the correct one :confused: